### ISSUE TITLE
Include dates of each training step

### ISSUE THREAD

**Issue by maneesha** [carpentries/amy#2477]
When a trainee looks for their training progress status, they see something like this:
![snippet of screenshot of trainee checkout status](https://github.com/carpentries/amy/assets/829690/4dc9d91f-b9a6-43f5-aaac-a4f98e55394d)

Language such as "passed" should be updated according to #2476.
The status should also include a date (i.e. "Welcome Session complete as of 10 July 2023." 

For training, this should be the actual training date (not the training passed recorded date).  For demo and welcome, this should be the recorded date because we may not have the event date. For get involved, this should be the date the submission was approved (not the date it was submitted).

**Comment by elichad** [comment]
If a trainee has multiple training steps, for example, should we use state/date of the most recent one?

I assume that any passed step will be the most recent step of its type, but there are probably edge cases where this isn't true (e.g. someone who never completed checkout re-takes training after some years and is asked to repeat)

**Comment by maneesha** [comment]
Tagging @karenword  and @klbarnes20  for feedback.

**Comment by klbarnes20** [comment]
I believe the most recent passed step has been the criteria @maneesha has been using in queries and that makes the most sense to me from an aggregated data perspective. For the trainee, maybe it would make more sense to show all passed? it that possible?

**Comment by karenword** [comment]
For training events, it may be desirable to show all events, but that doesn't help with an automated determination of whether the step was passed. This is maybe a more complex problem than it seems because a) old trainings that have been passed do eventually expire and have to be re-taken, but we do not currently have a rigid cutoff for this - it depends on engagement during the interim and b) we have had weird instances of people re-taking the training when they don't need to, and then ending up with a fail or ask-to-repeat because they don't stay for the full event. 

I think the ideal situation would be to display the date of the most-recent-passed event *conditional that the date is within the last 1.5 years*. (or date in past year with manual override option to allow an older event to count)

**Comment by elichad** [comment]
Given this feedback, would you be happy to use the following approach for all steps?
1. If there are one or more passed entries, use the most recent of those
2. If there are no passed entries, use the most recent non-passed entry

For training, I'm wary of hiding progress even if it's expired - it could be confusing for a trainee to look at the progress page and see "not completed yet" if they know they completed training some years before (and just showing the passed status with no date doesn't solve the problem at all). Also, this progress page remains available to everyone who's gone through IT, even if they now have their badge, so we do need to have older dates visible in that case.

We could instead show a note in the training step where applicable like "As you passed training more than <_a year_> ago but have not yet completed checkout, you may need to re-complete training to be eligible for instructor certification. Please contact <_email_> to discuss." This also works better with our vague boundaries for cutoff. 

After someone contacts us and we confirm the need to repeat, we could update that training progress state (and any other completed checkout steps?) to "Asked to repeat," and then any new progress would automatically be shown according to the rules above. (Or we could look at adding a new 'expired' state, but I think we would need to define specific cutoff criteria before using that in earnest, as we'd probably want to automate it)

What are your thoughts on this?

In short, it's _possible_ to build very complex logic into the page display, but it feels like a [code smell](https://en.wikipedia.org/wiki/Code_smell) to me. It would be better to handle things within the database records as much as possible.

**Comment by klbarnes20** [comment]
@elichad what you have suggested makes sense to me.
